### ISSUE TITLE
Different workers always process the same batch

### ISSUE THREAD

**Issue by MiguelMonteiro** [GoogleCloudPlatform/cloudml-samples#58]
For the census example in the model.input_fn() function for both [tf.train.shuffle_batch()](https://www.tensorflow.org/api_docs/python/tf/train/shuffle_batch) and [tf.train.batch()](https://www.tensorflow.org/api_docs/python/tf/train/batch) unless shared_name is specified both these queues are replicated for all workers and the master. 

This causes all workers to process all batches in the same order as every other worker (i.e. the same batch at the same time if they are going at the same speed). I don't know if this is intended behaviour, I expected each worker to process a new batch each time they took one from the queue.
 
The code I am referring to is under /census/tensorflowcore/trainer/model.py and the function is model.input_fn(), the specific bit is:

```
  if shuffle:
    features = tf.train.shuffle_batch(
        features,
        batch_size,
        min_after_dequeue=2 * batch_size + 1,
        capacity=batch_size * 10,
        num_threads=multiprocessing.cpu_count(),
        enqueue_many=True,
        allow_smaller_final_batch=True
    )
  else:
    features = tf.train.batch(
        features,
        batch_size,
        capacity=batch_size * 10,
        num_threads=multiprocessing.cpu_count(),
        enqueue_many=True,
        allow_smaller_final_batch=True
    )
```

**Comment by elibixby** [comment]
There's a performance trade-off to be had here (which doesn't really matter much as the census dataset is so small, but from a pedagogical perspective...). If `shared_name` is specified input data is must traverse the network several additional times: read as binary to the worker -> enqueued to variable on a param-server -> read as binary to worker -> enqueued as parsed tensor to param server -> read as parsed tensor to worker

Making the memory (for buffering) and network requirements of the cluster much larger. Now in this case, a single census example is quite small so you are probably right, we should be using global variables for the queues, but for networks with large features (like image networks), this would normally be the right approach.

**Comment by MiguelMonteiro** [comment]
Thanks for the response, I just wanted to be sure this is how it was working, I am making my own image network based on this code and I want all workers to pitch in for one epoch because the processing is very heavy (3D images). I am using a shared queue with filenames (each file is an image) instead of examples precisely to address the issue you have mentioned.

**Comment by elibixby** [comment]
@MiguelMonteiro if you find yourself performance constrained, you can also use a shared queue for the filenames, and a local queue for the shuffle_batcher, and split up your epoch between `num_worker` files, then only filenames will traverse the network, but workers will all work on different examples.

**Comment by elibixby** [comment]
Hahah, Didn't see your earlier comment. Looks like you are already doing exactly what I suggested =).

**Comment by elibixby** [comment]
@puneith I think for the move into census we should use shared_queue=True most likely.

**Comment by puneith** [comment]
@MiguelMonteiro This code uses tf.datasets in TF1.4 now. Please see https://www.tensorflow.org/api_docs/python/tf/data/TextLineDataset#shard and use that appropriately in the code. Feel free to send a PR as well as I see this as a generic sample feature.
